source/simulator/solver_schemes.cc

```diff
@@ -431,12 +431,19 @@ namespace aspect
         compute_current_constraints ();
       }
 
-    // the Stokes matrix depends on the viscosity. if the viscosity
-    // depends on other solution variables, then after we need to
-    // update the Stokes matrix in every time step and so need to set
-    // the following flag. if we change the Stokes matrix we also
-    // need to update the Stokes preconditioner.
-    rebuild_stokes_matrix = rebuild_stokes_preconditioner = assemble_newton_stokes_matrix = true;
+    // If the Stokes matrix depends on the solution, or we have active
+    // velocity boundary conditions, we need to re-assemble the system matrix
+    // (and preconditioner) every time. If we have active boundary conditions,
+    // they could a) depend on the solution, or b) be inhomogeneous. In both
+    // cases, just assembling the RHS will be incorrect.  If no active
+    // boundaries exist, we only have no-slip or free slip conditions, so we
+    // don't need to force assembly of the matrix.
+    if ((nonlinear_iteration == 0 && time_step == 0)
```

Comment:
Are you sure you need the first check here? I would think that the flags are true when the simulator starts? Also, isn't ``time_step`` the timestep size? Did you want to use ``time_step_number``?

Reply:
Yes, it should have been `time_step_number`. I had to add it to make it work, but I will test whether that is still the case.

Reply:
It works without it, so I removed it.